### PR TITLE
feat(admin): Layer detail page

### PR DESCRIPTION
**Pull request by corinacioloca**
By submitting a PR to this repository, you agree to the terms within the [Code of Conduct](https://github.com/natgeosociety/marapp-frontend/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/natgeosociety/marapp-frontend/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

> Describe the purpose of this PR along with any background information and the impacts of the proposed change. For the benefit of the community, please do not assume prior context.
>
> Provide details that support your chosen implementation, including: breaking changes, alternatives considered, changes to the API, etc.

### References

> Include any links supporting this change such as a:
>
> - GitHub Issue/PR number addressed or fixed
> - StackOverflow post
> - Support forum thread
> - Related pull requests/issues from other repos
>
> If there are no references, simply delete this section.

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in /docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not `master`

### CODE REVIEW

File: packages/earth-admin/src/components/detail-list/DetailList.tsx [hunk 1]
```diff
@@ -20,17 +20,26 @@
 import * as React from 'react';
 
 import { LinkWithOrg } from 'components/link-with-org';
-import { PlaceIntersectionProps } from 'pages-client/places/model';
+import classNames from 'classnames';
```
Comment (ancashoria) [inline packages/earth-admin/src/components/detail-list/DetailList.tsx]: Please move external imports at the top

File: packages/earth-admin/src/components/detail-list/DetailList.tsx [hunk 2]
```diff
@@ -20,17 +20,26 @@
...
+
+interface DetailListProps {
+  data: any,
```
Comment (ancashoria) [inline packages/earth-admin/src/components/detail-list/DetailList.tsx]: `data` should have stricter typing
```suggestion
interface ISomething {
  id: string
  name: string
}
  data: [ISomething],
```

File: packages/earth-admin/src/components/detail-list/DetailList.tsx [hunk 3]
```diff
@@ -20,17 +20,26 @@
...
+  key?: string | number;
+}
 
-export default function Intersections( props: PlaceIntersectionProps) {
-  const { intersections, name } = props;
+export default function DetailList(props: DetailListProps) {
+  const {data, name, type, className} = props;
 
   return (
     <div className="marapp-qa-placesintersections ng-flex ng-flex-column ng-margin-medium-bottom">
```
Comment (ancashoria) [inline packages/earth-admin/src/components/detail-list/DetailList.tsx]: The component is used in multiple locations, `marapp-qa-placesintersections` should be renamed to something more generic to reflect that.
I checked with qa and the class is not currently in use

File: packages/earth-admin/src/utils/hooks.tsx [hunk 4]
```diff
@@ -102,3 +102,22 @@ export function useDomWatcher(ref, callback, skip) {
...
+  ref.current.select();
+
+  document.execCommand('copy');
+  e.target.focus();
+  successFunction('Copied!');
+
+  setTimeout(() => {
+    successFunction('');
+  }, 4000);
+}
```
Comment (ancashoria) [inline packages/earth-admin/src/utils/hooks.tsx]: This doesn't look like a hook, we should move it to utils/index
Also, it needs types for it's params and it's return.

File: packages/earth-admin/src/utils/index.ts [hunk 5]
```diff
@@ -186,4 +186,35 @@ export const downloadFile = (data) => {
...
   const blobUrl = URL.createObjectURL(jsonBlob);
   return blobUrl;
-}
+};
+
+
+/**
+ * Flattens object array returned from multiselect to work with api
+ * @param data
+ * @param fieldName
```
Comment (ancashoria) [inline packages/earth-admin/src/utils/index.ts:196]: I usually skip this as the type information should be captured in TS types

File: packages/earth-admin/src/utils/index.ts [hunk 6]
```diff
@@ -186,4 +186,35 @@ export const downloadFile = (data) => {
...
+export const flattenObjectForSelect = (data: {}, fieldName: string) => {
+  return !!data ? data[fieldName] : data;
+};
+
+/**
+ * Sets up object needed for multiselect based on value returned from APIs
+ * @param options
+ * @param values
+ */
+export const getSelectValues = (options: {value: string}[], values) => {
```
Comment (ancashoria) [inline packages/earth-admin/src/utils/index.ts]: Don't forget to type the return value for all these functions.

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 7]
```diff
@@ -14,133 +14,476 @@
...
+import renderHTML from 'react-render-html';
+import Select from 'react-select';
+import Collapse from '@kunukn/react-collapse';
+import { Controller, useForm } from 'react-hook-form';
+
+import { encodeQueryToURL, formatDate, flattenArrayForSelect, flattenObjectForSelect, getSelectValues } from 'utils';
+import { getAllLayers, getLayer, handleLayerForm } from 'services/layers';
+import { useRequest, copyToClipboard } from 'utils/hooks';
+import classnames from 'classnames';
+import { navigate } from 'gatsby';
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]: Move external imports at the top

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 8]
```diff
@@ -14,133 +14,476 @@
...
+    ...{group: selectedGroup},
+  });
+  const {isLoading, data} = useRequest(() => getLayer(encodedQuery), {
+    permissions: AuthzGuards.writeLayersGuard,
+    query: encodedQuery,
+  });
+
+  const [layer, setLayer] = useState(data);
+  const [showDeleteModal, setShowDeleteModal] = useState(false);
+  const [formValid, setFormValid] = useState(true);
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]: This state seems to be a duplicate of `formState.isValid`. Why is that?

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 9]
```diff
@@ -14,133 +14,476 @@
...
+      ...(type && {type: flattenObjectForSelect(type, 'value')}),
+      ...(provider && {provider: flattenObjectForSelect(provider, 'value')}),
+      ...(references && {references: flattenArrayForSelect(references, 'id')}),
+    };
 
-      setIsLoading(false);
+    try {
+      setIsLoading && setIsLoading(true);
+      await handleLayerForm(false, parsed, id, selectedGroup);
+      const res = await getLayer(encodedQuery);
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx:144]: Is the extra get request needed? `handleLayerForm` seems to return the exact same data

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 10]
```diff
@@ -14,133 +14,476 @@
...
 
 
-export function LayerDetail(path: any) {
-  const {selectedGroup} = useAuth0();
-  const encodedQuery = encodeQueryToURL(`layers/${path.page}`, {
-    ...LAYER_DETAIL_QUERY,
-    ...{group: selectedGroup},
-  });
+  function handleServerErrors(errors) {
+    setServerErrors(errors);
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]: `serverErrors` is not being rendered anywhere. Not sure where to put it, but it should at least exist

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 11]
```diff
@@ -14,133 +14,476 @@
...
+              </InlineEditCard>
+            </div>
+            <div className="ng-width-1-4">
+              <Card>
+                <Toggle
+                  name="primary"
+                  label="Primary"
+                  value={primary}
+                  className="ng-display-block"
+                  onChange={(e) => onSubmit(e)}
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]:
```suggestion
                  onChange={onSubmit}
```

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 12]
```diff
@@ -14,133 +14,476 @@
...
+                  <p className="ng-margin-remove ng-padding-left">{type}</p>
+                </div>
+              </InlineEditCard>
+            </div>
+          </div>
+          <div className="ng-grid">
+            <div className="ng-width-1-1">
+              <InlineEditCard
+                onSubmit={onSubmit}
+                validForm={true}
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx:382]:
```suggestion
                validForm={formValid}
```

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 13]
```diff
@@ -14,133 +14,476 @@
...
-  return (
-    <ContentLayout errors={errors} backTo="/layers" isLoading={isLoading} className="marapp-qa-layersedit">
-      <LayerEdit data={data} newLayer={path.newLayer}/>
+                    <div className="ng-margin-medium-bottom">
+                      <label htmlFor="config">Layer Config</label>
+                      <Controller
+                        name="config"
+                        control={control}
+                        defaultValue={layerConfig}
+                        onChange={(layerConfig) => handleJsonChange(layerConfig)}
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]:
```suggestion
                        onChange={handleJsonChange}
```

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 14]
```diff
@@ -14,133 +14,476 @@
...
+                        as={<JsonEditor json={layerConfig}/>}
+                      />
+                    </div>
+
+                  </>
+                )}>
+                <div className="ng-margin-medium-bottom">
+                  {config && <div>
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]: Just to make it more readable
```suggestion
                  {config && (
                      <div>
                      ...
                      </div>
                   )}
```

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 15]
```diff
@@ -14,133 +14,476 @@
...
+                    <p className="ng-flex">
+                      <span className="ng-text-weight-medium">Layer config</span>
+                      <span>
+                        <i onClick={(e) => copyToClipboard(e, textAreaRef, setCopySuccess)}
+                           className='ng-icon ng-icon-layers ng-c-cursor-pointer ng-margin-small-horizontal'/>
+                        <span className="ng-text-weight-normal">{copySuccess}</span>
+                      </span>
+                    </p>
+
+                    <Collapse addState isOpen={toggle} collapseHeight="100px">
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]: `toggle` should be `toggleJSONEditor` to make it easier to understand

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 16]
```diff
@@ -14,133 +14,476 @@
...
+                    </div>
+                  </>
+                )}>
+                <div className="ng-margin-medium-bottom">
+                  {!!references ? <DetailList data={references} name='Layer References' type='layers'
+                                              className="ng-flex-column ng-flex-top"/> :
+                    <div>
+                      <p className="ng-text-weight-bold ng-margin-small-bottom">Layer references</p>
+                      <span className="ng-padding-left">No layer references</span></div>}
+                </div>
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx]: This needs some help to make it more readable. Consider () and line breaks

File: packages/earth-admin/src/pages-client/layers/details.tsx [hunk 17]
```diff
@@ -14,133 +14,476 @@
...
+              </InlineEditCard>
+
+            </div>
+          </div>
+        </form>
+
+        <input type="text" ref={textAreaRef} value={JSON.stringify(config)} readOnly={true}
```
Comment (ancashoria) [inline packages/earth-admin/src/pages-client/layers/details.tsx:477]: What's this?

File: packages/earth-admin/src/pages-client/layers/details.tsx (hunk 9 above)
Comment (corinacioloca) [reply packages/earth-admin/src/pages-client/layers/details.tsx:144]: put request doesn't bring back included fields

**Review by ancashoria — APPROVED** [review]

**Pull request merged**